### PR TITLE
ci(metrics): Run measurement functions in parallel

### LINKED ISSUE #14853: Make metrics parallel
If I understand correctly, we run `analysis-stats` in a single thread, so we can probably run metrics for different repositories in separate processes to make it faster (it currently takes ~50 minutes, but `analysis-stats` on self is ~15 minute).

### PR DESCRIPTION
**Pull request by NanthR**
Resolves #14853

### CODE REVIEW

File: xtask/src/flags.rs [hunk 1]
```diff
@@ -105,9 +107,32 @@ pub struct PublishReleaseNotes {
     pub dry_run: bool,
 }
 
+#[derive(Debug)]
+pub enum MeasurementType {
+    Build,
+    AnalyseSelf,
+    AnalyseRipgrep,
+    AnalyseWebRender,
+    AnalyseDiesel,
```
Comment (mattbonnell) [inline xtask/src/flags.rs:116]: nit: 'Analyze'

File: xtask/src/flags.rs [hunk 2]
```diff
@@ -42,7 +44,7 @@ xflags::xflags! {
             required changelog: String
         }
         cmd metrics {
-            optional --dry-run
```
Comment (mattbonnell) [inline xtask/src/flags.rs:45]: Should we be removing this?

**Comment by HKalbasi** [comment]
The implementation looks good to me. Is there a way to test it before merge?

File: .github/workflows/metrics.yaml [hunk 3]
```diff
@@ -11,20 +11,234 @@ env:
...
+        name: webrender-${{ github.sha }}
+
+    - name: Download diesel metrics
+      uses: actions/download-artifact@v3
+      with:
+        name: diesel-${{ github.sha }}
+
+    - name: Combine json
+      run: |
+        git clone --depth 1 https://$METRICS_TOKEN@github.com/nanthR/metrics.git
```
Comment (lnicola) [inline .github/workflows/metrics.yaml:238]: We should change this before merging.

File: .github/workflows/metrics.yaml [hunk 4]
```diff
@@ -11,20 +11,235 @@ env:
...
+        key: ${{ runner.os }}-cargo-${{ github.sha }}
+
+    - name: Restore target cache
+      uses: actions/cache@v3
+      with:
+        path: target/
+        key: ${{ runner.os }}-target-${{ github.sha }}
+
+    - name: Collect build metrics
+      run: cargo xtask metrics ripgrep
```
Comment (bjorn3) [inline .github/workflows/metrics.yaml:125]: Maybe use a build matrix to avoid all this duplication?

File: .github/workflows/metrics.yaml [hunk 5]
```diff
@@ -11,20 +11,235 @@ env:
   RUSTUP_MAX_RETRIES: 10
 
 jobs:
-  metrics:
-    if: github.repository == 'rust-lang/rust-analyzer'
+  setup_cargo:
     runs-on: ubuntu-latest
```
Comment (HKalbasi) [inline .github/workflows/metrics.yaml:15]: I assume this was removed for testing?

**Comment by HKalbasi** [comment]
@NanthR it failed, can you take a look? https://github.com/rust-lang/rust-analyzer/actions/runs/5402910351/jobs/9815765491

File: .github/workflows/metrics.yaml [hunk 6]
```diff
@@ -11,20 +11,135 @@ env:
...
+        name: diesel-${{ github.sha }}
+
+    - name: Combine json
+      run: |
+        git clone --depth 1 https://$METRICS_TOKEN@github.com/rust-analyzer/metrics.git
+        jq -s ".[0] * .[1] * .[2] * .[3] * .[4]" build.json self.json ripgrep.json webrender.json diesel.json -c >> metrics/metrics.json
+        cd metrics
+        git add .
+        git -c user.name=Bot -c user.email=dummy@example.com commit --message 📈
+        git push origin new
```
Comment (lowr) [inline .github/workflows/metrics.yaml:143]: Should this be `master` instead of `new`?

**Pull request merged**